{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[nltk_data] Downloading package wordnet to\n",
      "[nltk_data]     C:\\Users\\sarap\\AppData\\Roaming\\nltk_data...\n",
      "[nltk_data]   Package wordnet is already up-to-date!\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import nltk\n",
    "nltk.download('wordnet')\n",
    "from nltk.corpus import stopwords\n",
    "from nltk.stem import WordNetLemmatizer\n",
    "import re\n",
    "import string\n",
    "\n",
    "from gensim import corpora, models, similarities"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Finding the best matching research article entries for every patient profile\n",
    "\n",
    "On the first glance, the problem seems to be the simple case of similarity calculation between patient profiles and relevant research articles. However, some features like, \n",
    "\n",
    "+ prescribed drug\n",
    "+ genetic variants (rsIDs)\n",
    "\n",
    "have more contribution towards final decision than features like,\n",
    "\n",
    "+ diagnosis\n",
    "+ anscestry\n",
    "+ age\n",
    "    \n",
    "In case of supervised learning, the feature relevance (impact) calculation becomes simpler due to available techniques. This is more of a clustering problem, so normal feature relevance calculation is a moot point. Hence, the decision is based on the logic.\n",
    "\n",
    "```\n",
    "The research article entries could have hundreds of senarios with caucasian population suffering from type 2 diabetes. However, relevance of these entries will boost significantly if they are discussing the specific drug which is prescribed to the patient and also one of the genetic variants of the patient.\n",
    "```\n",
    "\n",
    "Hence, final decision was to build ensemble technique which primarily considers three features :\n",
    "1. prescribed drug\n",
    "2. genetic variants (rsIDs)\n",
    "3. similarity measure based on diagnostic_information and generic_information (age, sex, ancestry)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>Age</th>\n",
       "      <th>Sex</th>\n",
       "      <th>Ancestry</th>\n",
       "      <th>Diagnosis/Conditions</th>\n",
       "      <th>Drugs taken</th>\n",
       "      <th>rsIDs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>37</td>\n",
       "      <td>male</td>\n",
       "      <td>caucasian</td>\n",
       "      <td>HIV infection, type 2 diabetes</td>\n",
       "      <td>metformin</td>\n",
       "      <td>rs10306114,rs1042522,rs10455872,rs1045642,rs10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B</td>\n",
       "      <td>43</td>\n",
       "      <td>male</td>\n",
       "      <td>black/african american</td>\n",
       "      <td>epilepsy, seizures, stroke, nerve pain, asthma</td>\n",
       "      <td>metaproterenol</td>\n",
       "      <td>rs10306114,rs1042522,rs10455872,rs1045642,rs10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>54</td>\n",
       "      <td>female</td>\n",
       "      <td>asian</td>\n",
       "      <td>ulcerative colitis,inflammatory bowel disease,...</td>\n",
       "      <td>paroxetine</td>\n",
       "      <td>rs10306114,rs1042522,rs10455872,rs1045642,rs10...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Patient  Age     Sex                Ancestry  \\\n",
       "2       C   37    male               caucasian   \n",
       "1       B   43    male  black/african american   \n",
       "0      A    54  female                   asian   \n",
       "\n",
       "                                Diagnosis/Conditions     Drugs taken  \\\n",
       "2                     HIV infection, type 2 diabetes       metformin   \n",
       "1     epilepsy, seizures, stroke, nerve pain, asthma  metaproterenol   \n",
       "0  ulcerative colitis,inflammatory bowel disease,...      paroxetine   \n",
       "\n",
       "                                               rsIDs  \n",
       "2  rs10306114,rs1042522,rs10455872,rs1045642,rs10...  \n",
       "1  rs10306114,rs1042522,rs10455872,rs1045642,rs10...  \n",
       "0  rs10306114,rs1042522,rs10455872,rs1045642,rs10...  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "patient_profile = 'D:/MedicalNLP/job-post-ai-patient-profiles.csv'\n",
    "pat_prof_df = pd.read_csv(patient_profile, sep = '\\t')\n",
    "pat_prof_df.sample(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Variant</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Chemical</th>\n",
       "      <th>Phenotype Category</th>\n",
       "      <th>Significance</th>\n",
       "      <th>Notes</th>\n",
       "      <th>Sentence</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Article Title</th>\n",
       "      <th>Article Abstract</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>64</th>\n",
       "      <td>rs2032582</td>\n",
       "      <td>ABCB1 (PA267)</td>\n",
       "      <td>tacrolimus (PA451578)</td>\n",
       "      <td>metabolism/PK</td>\n",
       "      <td>yes</td>\n",
       "      <td>3 - 12 months post-transplant. In multivariate...</td>\n",
       "      <td>Genotype AA is associated with decreased clear...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>Relationship of CYP3A5 genotype and ABCB1 dipl...</td>\n",
       "      <td>Tacrolimus (TAC) is one of the most successful...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>rs1045642</td>\n",
       "      <td>ABCB1 (PA267)</td>\n",
       "      <td>midazolam (PA450496)</td>\n",
       "      <td>other,\"metabolism/PK\"</td>\n",
       "      <td>no</td>\n",
       "      <td>It's not clear exactly what genotype compariso...</td>\n",
       "      <td>Allele A is not associated with increased clea...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>Induction of CYP3A4 by vinblastine: Role of th...</td>\n",
       "      <td>Several microtubule targeting agents are capab...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>68</th>\n",
       "      <td>rs1045642</td>\n",
       "      <td>ABCB1 (PA267)</td>\n",
       "      <td>cyclosporine (PA449167)</td>\n",
       "      <td>efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>The A allele was found to have a higher freque...</td>\n",
       "      <td>Allele A is associated with decreased response...</td>\n",
       "      <td>chr7</td>\n",
       "      <td>A pharmacogenetic study of ABCB1 polymorphisms...</td>\n",
       "      <td>Psoriasis affects 2-3% of the population, caus...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>752</th>\n",
       "      <td>rs1800629</td>\n",
       "      <td>TNF (PA435)</td>\n",
       "      <td>infliximab (PA452639)</td>\n",
       "      <td>efficacy</td>\n",
       "      <td>yes</td>\n",
       "      <td>Good responders defined as patients whose dise...</td>\n",
       "      <td>Genotype GG is associated with increased respo...</td>\n",
       "      <td>chr6</td>\n",
       "      <td>Polymorphism at position -308 of the tumor nec...</td>\n",
       "      <td>To test whether the G-to-A polymorphism at pos...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1578</th>\n",
       "      <td>rs4149056</td>\n",
       "      <td>SLCO1B1 (PA134865839)</td>\n",
       "      <td>atorvastatin (PA448500),\"simvastatin (PA451363)\"</td>\n",
       "      <td>efficacy</td>\n",
       "      <td>no</td>\n",
       "      <td>While, on average, total cholesterol, triglyce...</td>\n",
       "      <td>Genotype CC is not associated with increased r...</td>\n",
       "      <td>chr12</td>\n",
       "      <td>Lack of association between SLCO1B1 polymorphi...</td>\n",
       "      <td>There is significant inter-individual variabil...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Variant                   Gene  \\\n",
       "64    rs2032582          ABCB1 (PA267)   \n",
       "27    rs1045642          ABCB1 (PA267)   \n",
       "68    rs1045642          ABCB1 (PA267)   \n",
       "752   rs1800629            TNF (PA435)   \n",
       "1578  rs4149056  SLCO1B1 (PA134865839)   \n",
       "\n",
       "                                              Chemical     Phenotype Category  \\\n",
       "64                               tacrolimus (PA451578)          metabolism/PK   \n",
       "27                                midazolam (PA450496)  other,\"metabolism/PK\"   \n",
       "68                             cyclosporine (PA449167)               efficacy   \n",
       "752                              infliximab (PA452639)               efficacy   \n",
       "1578  atorvastatin (PA448500),\"simvastatin (PA451363)\"               efficacy   \n",
       "\n",
       "     Significance                                              Notes  \\\n",
       "64            yes  3 - 12 months post-transplant. In multivariate...   \n",
       "27             no  It's not clear exactly what genotype compariso...   \n",
       "68            yes  The A allele was found to have a higher freque...   \n",
       "752           yes  Good responders defined as patients whose dise...   \n",
       "1578           no  While, on average, total cholesterol, triglyce...   \n",
       "\n",
       "                                               Sentence Chromosome  \\\n",
       "64    Genotype AA is associated with decreased clear...       chr7   \n",
       "27    Allele A is not associated with increased clea...       chr7   \n",
       "68    Allele A is associated with decreased response...       chr7   \n",
       "752   Genotype GG is associated with increased respo...       chr6   \n",
       "1578  Genotype CC is not associated with increased r...      chr12   \n",
       "\n",
       "                                          Article Title  \\\n",
       "64    Relationship of CYP3A5 genotype and ABCB1 dipl...   \n",
       "27    Induction of CYP3A4 by vinblastine: Role of th...   \n",
       "68    A pharmacogenetic study of ABCB1 polymorphisms...   \n",
       "752   Polymorphism at position -308 of the tumor nec...   \n",
       "1578  Lack of association between SLCO1B1 polymorphi...   \n",
       "\n",
       "                                       Article Abstract  \n",
       "64    Tacrolimus (TAC) is one of the most successful...  \n",
       "27    Several microtubule targeting agents are capab...  \n",
       "68    Psoriasis affects 2-3% of the population, caus...  \n",
       "752   To test whether the G-to-A polymorphism at pos...  \n",
       "1578  There is significant inter-individual variabil...  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "research_info = 'D:/MedicalNLP/job-post-ai-database-results.csv'\n",
    "research_df = pd.read_csv(research_info, sep = '\\t')\n",
    "research_df.sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Readying the data\n",
    "\n",
    "\n",
    "### Feature combination and elimination\n",
    "#### Patient profiles\n",
    "All the information given here is precise and significant. Hence, no elimination was performed. Although, feature combination was done combining features depicting diagnostic_information and generic_information (age, sex, ancestry).\n",
    "#### Research articles\n",
    "There is overwhelming amount of irrelevant information in this dataset. The patient profile does not discuss features like,\n",
    "+ gene\n",
    "+ chromosome\n",
    "\n",
    "Also following features can be considered later while studying the relevant articles, as our similarity calculation is not semantic but BOW based\n",
    "+ phenotype_catagory\n",
    "+ significance \n",
    "\n",
    "Additional the feature, \n",
    "+ article abstract\n",
    "\n",
    "consists of too much information containing lengthy genetic sequences, numeric values, units of measures (mg/DL, % etc.). In absence of sophisticated information extraction, this field worsens the final hypothesis. Because it adds a lot of irrelevant vocabulary in our BOW. \n",
    "Hence all these features are eliminated.\n",
    "\n",
    "### Data clean-up\n",
    "\n",
    "The data preparation / data clean-up is the crucial step when building ML solution. Especially, for NLP it becomes a huge responsibility as there is no fixed set of dos and don'ts. If one is using dictionary-based methods, it is natural to prune the data removing any irrelevant numbers, sequences, stop words, punctuations. But employing these techniques will be foolish when one used word embeddings or encoder (Neural Networks) based techniques, as everything hinges upon relative positions in the sentence and around other words. \n",
    "\n",
    "For this case, I decided to use traditional ML (BOW + tfidf) and gensim for relevance calculation. Hence I am using data clean-up techniques like punctuation + stop-word removal, lemmatization for easier correlation etc.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# using traditional NLP techniques like case altering, punctuation{ stopword} removal and lemmatization\n",
    "# effective for BOW/dictionary based techniques but removes positional or relative information\n",
    "# not recommended for deep learning based techniques like word vectors or encoders\n",
    "stopset = stopwords.words('english')\n",
    "translator = str.maketrans(string.punctuation, ' '*len(string.punctuation)) \n",
    "wordnet_lemmatizer = WordNetLemmatizer()\n",
    "def clean_text(text_ex):\n",
    "\n",
    "    #print(text_ex)\n",
    "    #remove numbers but keep alphanumeric\n",
    "    #text_ex = re.sub(r'\\W+', ' ', text_ex)\n",
    "    \n",
    "    # removing digits\n",
    "    #text_ex = re.sub(r\"\\d\", \"\", text_ex)\n",
    "    \n",
    "    #removing single alphabets\n",
    "    text_ex = re.sub(r\"\\s+[a-zA-Z]\\s+\", \" \", text_ex)\n",
    "    #text_ex = re.sub(r\"\\s+[a-zA-Z]+[a-zA-Z]\\s+\", \" \", text_ex)\n",
    "    \n",
    "    text_ex = text_ex.lower()\n",
    "    \n",
    "    #map punctuation to space\n",
    "    text_ex = text_ex.translate(translator)\n",
    "    \n",
    "    tokens = nltk.word_tokenize(text_ex)\n",
    "    #stopword removal\n",
    "    tokens = [w for w in tokens if w not in stopset]\n",
    "    tokens = [wordnet_lemmatizer.lemmatize(w) for w in tokens]\n",
    "    \n",
    "    return \" \".join(tokens)\n",
    "\n",
    "def remove_regex(text_ex):\n",
    "\n",
    "    #removing regex\n",
    "    return re.sub(r\"\\spa\\d+\", \" \", text_ex)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Patient</th>\n",
       "      <th>profile</th>\n",
       "      <th>drug</th>\n",
       "      <th>gen_variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>B</td>\n",
       "      <td>adult male black african american epilepsy sei...</td>\n",
       "      <td>metaproterenol</td>\n",
       "      <td>rs10306114 rs1042522 rs10455872 rs1045642 rs10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>C</td>\n",
       "      <td>adult male caucasian hiv infection type 2 diab...</td>\n",
       "      <td>metformin</td>\n",
       "      <td>rs10306114 rs1042522 rs10455872 rs1045642 rs10...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>adult female asian ulcerative colitis inflamma...</td>\n",
       "      <td>paroxetine</td>\n",
       "      <td>rs10306114 rs1042522 rs10455872 rs1045642 rs10...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Patient                                            profile            drug  \\\n",
       "1       B  adult male black african american epilepsy sei...  metaproterenol   \n",
       "2       C  adult male caucasian hiv infection type 2 diab...       metformin   \n",
       "0      A   adult female asian ulcerative colitis inflamma...      paroxetine   \n",
       "\n",
       "                                         gen_variant  \n",
       "1  rs10306114 rs1042522 rs10455872 rs1045642 rs10...  \n",
       "2  rs10306114 rs1042522 rs10455872 rs1045642 rs10...  \n",
       "0  rs10306114 rs1042522 rs10455872 rs1045642 rs10...  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# data readying for patient profiles\n",
    "\n",
    "patient_profile = 'D:/MedicalNLP/job-post-ai-patient-profiles.csv'\n",
    "pat_prof_df = pd.read_csv(patient_profile, sep = '\\t')\n",
    "\n",
    "# finding specific age value in the data is challenging. However, found some articles discussing drug effects on adults and children\n",
    "# Hence converting numerical to catagorical variable\n",
    "pat_prof_df['Age'] = pat_prof_df['Age'].apply(lambda x: 'adult' if x>20 else 'child')\n",
    "\n",
    "#combining interesting columns\n",
    "pat_prof_df['profile'] = pat_prof_df.Age +\" \" + pat_prof_df.Sex + \" \" + pat_prof_df.Ancestry +\" \"+pat_prof_df['Diagnosis/Conditions'] + \" \" + pat_prof_df['Drugs taken']\n",
    "pat_prof_df['drug'] = pat_prof_df['Drugs taken'] \n",
    "pat_prof_df['gen_variant'] = pat_prof_df['rsIDs']\n",
    "\n",
    "# dropping insignificant data\n",
    "pat_prof_df = pat_prof_df.drop(['Age', 'Sex', 'Ancestry', 'Diagnosis/Conditions', 'rsIDs', 'Drugs taken'], axis=1)\n",
    "\n",
    "#removing punctuations, stopwords and lemmatizing the text\n",
    "pat_prof_df['profile'] = pat_prof_df['profile'].apply(clean_text)\n",
    "pat_prof_df['drug'] = pat_prof_df['drug'].apply(clean_text)\n",
    "pat_prof_df['gen_variant'] = pat_prof_df['gen_variant'].apply(clean_text)\n",
    "\n",
    "\n",
    "pat_prof_df.sample(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>drug</th>\n",
       "      <th>gen_variant</th>\n",
       "      <th>summary</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1935</th>\n",
       "      <td>warfarin</td>\n",
       "      <td>rs9923231</td>\n",
       "      <td>allele associated clearance warfarin compared ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>167</th>\n",
       "      <td>apixaban</td>\n",
       "      <td>rs1045642</td>\n",
       "      <td>genotype aa ag associated clearance apixaban p...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>carbamazepine</td>\n",
       "      <td>rs1045642</td>\n",
       "      <td>allele associated increased dose carbamazepine...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>790</th>\n",
       "      <td>folic acid  hydroxychloroquine  methotrexate  ...</td>\n",
       "      <td>rs1801133</td>\n",
       "      <td>allele associated response folic acid hydroxyc...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>857</th>\n",
       "      <td>tipifarnib</td>\n",
       "      <td>rs2032582</td>\n",
       "      <td>allele associated increased metabolism tipifar...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1880</th>\n",
       "      <td>warfarin</td>\n",
       "      <td>rs9923231</td>\n",
       "      <td>allele associated decreased dose warfarin chil...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>651</th>\n",
       "      <td>fentanyl</td>\n",
       "      <td>rs1799971</td>\n",
       "      <td>allele associated dose fentanyl people pain po...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>opioid anesthetic  general anesthetic  volatil...</td>\n",
       "      <td>rs1799971</td>\n",
       "      <td>genotype gg associated increased response opio...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1317</th>\n",
       "      <td>nevirapine</td>\n",
       "      <td>rs28399499</td>\n",
       "      <td>genotype cc ct associated decreased clearance ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>tacrolimus</td>\n",
       "      <td>rs1045642</td>\n",
       "      <td>allele associated dose adjusted trough concent...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                   drug gen_variant  \\\n",
       "1935                                          warfarin    rs9923231   \n",
       "167                                           apixaban    rs1045642   \n",
       "34                                       carbamazepine    rs1045642   \n",
       "790   folic acid  hydroxychloroquine  methotrexate  ...   rs1801133   \n",
       "857                                         tipifarnib    rs2032582   \n",
       "1880                                          warfarin    rs9923231   \n",
       "651                                           fentanyl    rs1799971   \n",
       "583   opioid anesthetic  general anesthetic  volatil...   rs1799971   \n",
       "1317                                        nevirapine   rs28399499   \n",
       "277                                         tacrolimus    rs1045642   \n",
       "\n",
       "                                                summary  \n",
       "1935  allele associated clearance warfarin compared ...  \n",
       "167   genotype aa ag associated clearance apixaban p...  \n",
       "34    allele associated increased dose carbamazepine...  \n",
       "790   allele associated response folic acid hydroxyc...  \n",
       "857   allele associated increased metabolism tipifar...  \n",
       "1880  allele associated decreased dose warfarin chil...  \n",
       "651   allele associated dose fentanyl people pain po...  \n",
       "583   genotype gg associated increased response opio...  \n",
       "1317  genotype cc ct associated decreased clearance ...  \n",
       "277   allele associated dose adjusted trough concent...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# data readying for research articles\n",
    "\n",
    "research_info = 'D:/MedicalNLP/job-post-ai-database-results.csv'\n",
    "research_df = pd.read_csv(research_info, sep = '\\t')\n",
    "\n",
    "# taking care of missing values by adding stopwords in-place which will be removed later\n",
    "research_df['Notes'] = research_df['Notes'].fillna('is of')\n",
    "research_df['Article Abstract'] = research_df['Article Abstract'].fillna('is of')\n",
    "\n",
    "#combining interesting columns\n",
    "research_df['drug'] = research_df['Chemical']\n",
    "research_df['gen_variant'] = research_df['Variant']\n",
    "research_df['summary'] = research_df['Sentence']+\" \" +  research_df['Article Title']+ \" \"+ research_df['Notes']\n",
    "\n",
    "\n",
    "# dropping insignificant data\n",
    "research_df = research_df.drop(['Gene', 'Variant','Chemical', 'Phenotype Category', 'Significance', 'Notes', 'Sentence', 'Chromosome', 'Article Title', 'Article Abstract'], axis=1)\n",
    "\n",
    "\n",
    "#removing punctuations, stopwords and lemmatizing the text\n",
    "research_df['drug'] = research_df['drug'].apply(clean_text)\n",
    "research_df['drug'] = research_df['drug'].apply(remove_regex)\n",
    "research_df['summary'] = research_df['summary'].apply(clean_text)\n",
    "research_df['gen_variant'] = research_df['gen_variant'].apply(clean_text)\n",
    "\n",
    "research_df.sample(10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Constructing the model\n",
    "\n",
    "\n",
    "### Model Selection\n",
    "#### Similarity Calculation\n",
    "Model to check if patient profile is relevant to corresponding research summary. After much pondering on which model to use for similarity calculation, I decided to do it with tfidf, gensim similarity measures. Besides considering that this is domain specific data, own vocabulary building using tfidf and BOW seems more sensible choice. .\n",
    "\n",
    "Gensim similarity calculation is much more optimized for sparse matrix than custom similarity (Cosine or Jaccard) comparison that I could write. I have more control over the similarity calculation than black box clustering like LDA or LSA. \n",
    "\n",
    "#### Justification for not going for the kill with Keras and LSTM\n",
    "+ Despite of the popular belief, I consider keeping the usage of LSTM based architectures as the last resort, especially for NLP. For most cases, traditional BOW techniques work fine according to my experience.\n",
    "\n",
    "+ Using pretrained embeddings (wordVector, fastText, gloVe) or encoder weights (BERT) is not really an option due to highly domain specific nature of the data.\n",
    "\n",
    "Note: Although I did find an article on clinicalBERT which might be of interest, if actual product needs to be designed and with no time constraints and relaxed research timeline\n",
    "https://towardsdatascience.com/how-do-they-apply-bert-in-the-clinical-domain-49113a51be50\n",
    "\n",
    "+ To train our own embeddings or encodings, it is expected to have large training set, which clearly, I don’t have.\n",
    "\n",
    "\n",
    "### Ensemble Building\n",
    "Like discussed earlier, decision behind building ensemble model originated from the logic that, features like, \n",
    "\n",
    "+ prescribed drug\n",
    "+ genetic variants (rsIDs)\n",
    "\n",
    "should have more contribution towards final decision than features like,\n",
    "\n",
    "+ diagnosis\n",
    "+ ancestry\n",
    "+ age\n",
    "\n",
    "which could be common for large number of people, but not all of them would take the prescribed drug.\n",
    "Hence, model ensemble primarily considers three features, first two being the important ones:\n",
    "1. prescribed drug\n",
    "2. genetic variants (rsIDs)\n",
    "3. similarity measure based on diagnostic_information and generic_information (age, sex, ancestry)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1167 1556  140  944 1310 1252  156   29   12 1650  736 1726  865   39\n",
      " 1725 1561  892  874  308  738 1248  920 1144  363  217  104  400 1759\n",
      "   97   98 1447  408  407 1039 1651  893  709 1654  364  164  403  622\n",
      " 1775  253  742  989  956  398  409  401]\n",
      "[0.07494742 0.07595357 0.07616989 0.07970971 0.0817286  0.08426116\n",
      " 0.08506846 0.08607369 0.08613807 0.14872275 0.13482411 0.12041564\n",
      " 0.08633952 0.12652117 0.11231668 0.09532007 0.12925307 0.1267301\n",
      " 0.09304031 0.15460253 0.09087431 0.14103013 0.0907613  0.16103905\n",
      " 0.09102055 0.09017017 0.1270077  0.1548422  0.13004579 0.10925093\n",
      " 0.09528142 0.11974282 0.12053156 0.09084714 0.12890793 0.09567098\n",
      " 0.176169   0.2367297  0.25122806 0.30487698 0.18405755 0.21298423\n",
      " 0.21543376 0.3026088  0.28115982 0.2998783  0.3027596  0.18765494\n",
      " 0.19303316 0.2339108 ]\n",
      "[1726, 39, 874, 920, 1651, 1775]\n",
      "[0.12041564 0.12652117 0.1267301  0.14103013 0.12890793 0.21543376]\n",
      "[ 877 1161  878  685 1122  109   83 1252 1235  217 1039 1310  200   73\n",
      " 1505 1447  852  851  836 1504 1228  348 1427 1561  838  839 1338  879\n",
      "  110  106 1415 1793 1792 1294 1296 1896 1791 1859 1765 1244 1174  834\n",
      " 1097  835 1330 1111 1930  107 1117 1780]\n",
      "[0.09485967 0.09527042 0.09537318 0.09849165 0.11927801 0.1510716\n",
      " 0.09541028 0.10249649 0.09827619 0.11071872 0.11050778 0.09941584\n",
      " 0.10169309 0.09964809 0.1110153  0.11590169 0.12045807 0.11268982\n",
      " 0.11183222 0.1032318  0.12045807 0.13342273 0.11614601 0.11594871\n",
      " 0.09736458 0.11234028 0.10183971 0.09547147 0.14550367 0.1512125\n",
      " 0.15751155 0.20505646 0.20505646 0.21502441 0.15350634 0.24041046\n",
      " 0.20505646 0.22241879 0.22241879 0.31209946 0.22784531 0.15966532\n",
      " 0.22241879 0.18543538 0.15702151 0.24690099 0.23991197 0.15743391\n",
      " 0.17282194 0.15543878]\n",
      "[]\n",
      "[]\n",
      "[1371 1872 1352 1431 1886 1468 1469  782 1247 1478 1298 1377 1456 1450\n",
      " 1357   15  518 1317 1521 1428 1336 1340  418  822  546  537 1394 1393\n",
      " 1167 1392 1389  639 1362 1458  409 1290 1294 1360 1355 1487 1351 1583\n",
      " 1590 1591 1592 1343 1795  554 1797 1796]\n",
      "[0.0759896  0.07714949 0.07749417 0.07826118 0.07908051 0.09569941\n",
      " 0.09592776 0.0881491  0.07911213 0.09504492 0.08186568 0.08275428\n",
      " 0.09324688 0.08426479 0.08185822 0.08275428 0.09407829 0.08261298\n",
      " 0.0920209  0.08614333 0.08318093 0.08426479 0.09404787 0.09597013\n",
      " 0.20045918 0.15860307 0.11935782 0.0989707  0.10468031 0.12898546\n",
      " 0.09974699 0.12110528 0.12050003 0.11940222 0.11034663 0.1307672\n",
      " 0.12137693 0.10267215 0.14511067 0.15096886 0.14030233 0.12253625\n",
      " 0.29259747 0.09930739 0.31825662 0.20609954 0.17101032 0.20139553\n",
      " 0.32892898 0.4396144 ]\n",
      "[1796]\n",
      "[0.4396144]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[[1726, 39, 874, 920, 1651, 1775], [], [1796]]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# building vocabulary for research articles\n",
    "summary_data = research_df['summary'].values.tolist() \n",
    "summary_dict = corpora.Dictionary(d.split() for d in summary_data)\n",
    "f2_cnt = len(summary_dict.token2id)\n",
    "summary_corpus = [summary_dict.doc2bow(row.split()) for row in summary_data]\n",
    "tfidf_summary = models.TfidfModel(summary_corpus)\n",
    "summary_index = similarities.SparseMatrixSimilarity(tfidf_summary[summary_corpus], num_features = f2_cnt)\n",
    "\n",
    "\n",
    "#Ensemble model building\n",
    "profile_relevant_research = []\n",
    "top_n = 50 # find top_n hypotheses besed on profile similairty\n",
    "for ind in pat_prof_df.index:\n",
    "    profile= pat_prof_df['profile'][ind]\n",
    "    profile = summary_dict.doc2bow(profile.split())\n",
    "    prof_sim = summary_index[tfidf_summary[profile]]\n",
    "\n",
    "    # take top n similar documents and then eliminate from them\n",
    "    top_similar = np.argpartition(prof_sim, -top_n)[-top_n:]\n",
    "    \n",
    "    print(top_similar)\n",
    "    print(prof_sim[top_similar])\n",
    "    \n",
    "    drug = pat_prof_df['drug'][ind]\n",
    "    gen_variant = pat_prof_df['gen_variant'][ind]\n",
    "    top_similar_modified = []\n",
    "    for ind_2 in top_similar:\n",
    "        variant_flag = False\n",
    "        drug_flag = False\n",
    "        res_drug = research_df['drug'][ind_2]\n",
    "        res_gen_variant = research_df['gen_variant'][ind_2]\n",
    "        \n",
    "        # note instead of this one on one finding one could use more sophisticated NER software or \n",
    "        # even similarity calculation building new drug based BOW, see below\n",
    "        for var in gen_variant:\n",
    "            if  res_gen_variant == var:\n",
    "                variant_flag = True\n",
    "                \n",
    "        if drug in res_drug:\n",
    "            drug_flag = True\n",
    "        \n",
    "        # instead of making it conditional we can allot importance to 3 features (0.3, 0.3, 0.4) or (0.4, 0.4, 02)\n",
    "        # and calculate confidance score by softmaxing similarty_score*importance\n",
    "        if drug_flag or variant_flag:\n",
    "            top_similar_modified.append(ind_2)\n",
    "            \n",
    "    profile_relevant_research.append(top_similar_modified)\n",
    "         \n",
    "    print(top_similar_modified)\n",
    "    print(prof_sim[top_similar_modified])\n",
    "\n",
    "profile_relevant_research"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Patient profile :  adult female asian ulcerative colitis inflammatory bowel disease anxiety kidney transplant paroxetine\n",
      "Drug :  paroxetine\n",
      "Found 6 relevant research articles\n",
      "Relavant document \n",
      " allele associated response fluvoxamine paroxetine people depressive disorder major compared allele effect serotonin type 2a 3a 3b receptor serotonin transporter gene paroxetine fluvoxamine efficacy adverse drug reaction depressed japanese patient association found comparing responder v non responder\n",
      "Relavant document \n",
      " allele associated response paroxetine people depressive disorder major compared allele g pharmacogenetics modern psychiatry\n",
      "Relavant document \n",
      " allele associated response paroxetine people depressive disorder major compared allele c pharmacogenetics modern psychiatry\n",
      "Relavant document \n",
      " genotype cc ct associated decreased response paroxetine people depression compared genotype aa abcb1 mdr1 gene polymorphism associated clinical response paroxetine patient major depressive disorder tri allelic snp comparison done change ham score week 6 strand 18 cc v 31 ca ct v 19 aa tt\n",
      "Relavant document \n",
      " allele associated increased response paroxetine people depressive disorder major compared allele g catechol methyltransferase val 108 158 met polymorphism affect antidepressant response paroxetine naturalistic setting aa ag gg response met allele val effect became significant week 3 treatment\n",
      "Relavant document \n",
      " allele associated increased dose paroxetine people depressive disorder major compared allele c cyp1a2 genetic polymorphism associated treatment response antidepressant paroxetine dose required week 4 treatment\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Patient profile :  adult male black african american epilepsy seizure stroke nerve pain asthma metaproterenol\n",
      "Drug :  metaproterenol\n",
      "Found 0 relevant research articles\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Patient profile :  adult male caucasian hiv infection type 2 diabetes metformin\n",
      "Drug :  metformin\n",
      "Found 1 relevant research articles\n",
      "Relavant document \n",
      " allele associated increased response metformin compared allele c tcf7l2 genetic variation augments incretin resistance influence response sulfonylurea metformin study understand genetics acute response metformin glipizide human sugar mgh measured decrease fasting glucose compared pre treatment people risk factor type 2 diabetes\n",
      "\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Performance demo\n",
    "for ind in pat_prof_df.index:\n",
    "    print(\"Patient profile : \",pat_prof_df['profile'][ind], )\n",
    "    print(\"Drug : \",pat_prof_df['drug'][ind])\n",
    "    relevant_research = profile_relevant_research[ind]\n",
    "    print(\"Found %d relevant research articles\" % len(relevant_research))\n",
    "    for doc in relevant_research:\n",
    "        print(\"Relavant document \\n\", research_df['summary'][doc])\n",
    "    print(\"\\n\\n\\n\")\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1.0\n",
      "paroxetine research results  paroxetine \n",
      "paroxetine research results  paroxetine \n",
      "paroxetine research results  paroxetine \n",
      "paroxetine research results  paroxetine \n",
      "paroxetine research results  fluvoxamine  milnacipran  paroxetine \n",
      "paroxetine research results  citalopram  fluoxetine  paroxetine \n",
      "paroxetine research results  citalopram  fluoxetine  paroxetine \n",
      "paroxetine research results  fluvoxamine  paroxetine \n",
      "paroxetine research results  paroxetine \n",
      "paroxetine research results  citalopram  fluoxetine  paroxetine \n",
      "most similar text 0\n",
      "1997\n",
      "paroxetine\n",
      "latanoprost \n"
     ]
    }
   ],
   "source": [
    "# drug based BOW experiments\n",
    "drug_data = research_df['drug'].values.tolist() \n",
    "drug_dict = corpora.Dictionary(d.split() for d in drug_data)\n",
    "f1_cnt = len(drug_dict.token2id)\n",
    "drug_corpus = [drug_dict.doc2bow(row.split()) for row in drug_data]\n",
    "tfidf_drug = models.TfidfModel(drug_corpus)\n",
    "drug_index = similarities.SparseMatrixSimilarity(tfidf_drug[drug_corpus], num_features = f1_cnt)\n",
    "\n",
    "    \n",
    "text1 = pat_prof_df['drug'][0]\n",
    "\n",
    "new_vec = drug_dict.doc2bow(text1.split())\n",
    "\n",
    "\n",
    "sim = drug_index[tfidf_drug[new_vec]]\n",
    "print(max(sim))\n",
    "max_sim = 0\n",
    "for i in range(len(sim)):\n",
    "    if sim[i] > 0:\n",
    "        print(text1, \"research results \", research_df['drug'][i])\n",
    "print(\"most similar text\", max_sim)\n",
    "print(len(sim))\n",
    "print(text1)\n",
    "print(research_df['drug'][max_sim])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hiv infection type 2 diabetes metformin\n",
      "0.116386056\n",
      "genotype del del associated increased response benazepril perindopril people diabetes mellitus compared genotype atacagtcactttttttttttttttgagacggagtctcgctctgtcgccc atacagtcactttttttttttttttgagacggagtctcgctctgtcgccc atacagtcactttttttttttttttgagacggagtctcgctctgtcgccc del ace dd genotype susceptible ace ii id genotype antiproteinuric effect ace inhibitor patient proteinuric non insulin dependent diabetes mellitus patient del del genotype significantly greater percentage reduction baseline 3 month treatment urinary excretion protein albumin compared remaining genotype indicates greater response antiproteinuric effect benazepril perindopril\n",
      "0.11842383\n",
      "genotype tt associated decreased metabolism efavirenz people hiv infection compared genotype gt high plasma efavirenz concentration cyp2b6 polymorphism thai hiv 1 infection significantly higher mid dose efv plasma concentration 12 week seen patient tt genotype compared patient gt genotype\n",
      "0.118706554\n",
      "genotype gt associated decreased metabolism efavirenz people hiv infection compared genotype gg high plasma efavirenz concentration cyp2b6 polymorphism thai hiv 1 infection significantly higher mid dose efv plasma concentration 12 week seen patient gt genotype compared patient gg genotype\n",
      "0.17784373\n",
      "genotype tt associated increased concentration efavirenz people hiv infection compared genotype gg gt successful efavirenz dose reduction hiv type 1 infected individual cytochrome p450 2b6 6 26 data unavailable\n",
      "0.12343557\n",
      "genotype tt associated decreased metabolism efavirenz people hiv infection compared genotype gg high plasma efavirenz concentration cyp2b6 polymorphism thai hiv 1 infection significantly higher mid dose efv plasma concentration 12 week seen patient tt genotype compared patient gg genotype\n",
      "0.18135631\n",
      "genotype gt tt associated increased concentration efavirenz people hiv infection compared genotype gg effect cyp2b6 gene polymorphism efavirenz plasma concentration chinese patient hiv infection data unavailable\n",
      "0.19576344\n",
      "genotype del del associated decreased response captopril people diabetes mellitus type 2 compared genotype atacagtcactttttttttttttttgagacggagtctcgctctgtcgccc atacagtcactttttttttttttttgagacggagtctcgctctgtcgccc atacagtcactttttttttttttttgagacggagtctcgctctgtcgccc del impact renin angiotensin system polymorphism renal haemodynamic responsiveness acute angiotensin converting enzyme inhibition type 2 diabetes mellitus 99mtc mag3 clearance mag3cle value increased addition ata allele treatment captopril mag3cle provides estimation effective renal plasma flow erpf therefore responsiveness captopril treatment\n",
      "0.24772844\n",
      "genotype del del associated increased response losartan people diabetes mellitus type 2 compared genotype atacagtcactttttttttttttttgagacggagtctcgctctgtcgccc atacagtcactttttttttttttttgagacggagtctcgctctgtcgccc atacagtcactttttttttttttttgagacggagtctcgctctgtcgccc del ace gene polymorphism serum ace activity iranian type ii diabetic patient macroalbuminuria snp presented ace patient type ii diabetes mellitus also macroalbuminuria dd genotype better response losartan compared patient carrying allele normoalbuminuria\n",
      "0.3898213\n",
      "allele associated response pioglitazone people diabetes mellitus compared allele c cyp2c8 slco1b1 variant therapeutic response thiazolidinediones patient type 2 diabetes data unavailable\n",
      "0.24888565\n",
      "genotype del del associated increased response captopril people diabetes mellitus type 2 compared genotype atacagtcactttttttttttttttgagacggagtctcgctctgtcgccc atacagtcactttttttttttttttgagacggagtctcgctctgtcgccc atacagtcactttttttttttttttgagacggagtctcgctctgtcgccc del ace gene polymorphism serum ace activity iranian type ii diabetic patient macroalbuminuria snp presented ace patient type ii diabetes mellitus normoalbuminuric dd genotype better response captopril compared patient carrying allele macroalbuminuria\n",
      "0.2093833\n",
      "genotype tt associated decreased response sulfonamide urea derivative people diabetes mellitus type 2 compared genotype cc variation tcf7l2 influence therapeutic response sulfonylurea godarts study data unavailable\n",
      "0.25464234\n",
      "allele associated response drug used diabetes people diabetes mellitus type 2 compared allele c pharmacogenetics oral antidiabetes drug evidence diverse signal irs1 locus case patient requiring insulin therapy added either instead maximal near maximal dos oral antidiabetes drug oads control patient acceptable glycemic control hba1c 8 absence insulin therapy oads mostly sulfonylurea metformin 2 4 patient also thiazolidinediones gliptins\n",
      "0.54391336\n",
      "allele associated increased response metformin compared allele c tcf7l2 genetic variation augments incretin resistance influence response sulfonylurea metformin study understand genetics acute response metformin glipizide human sugar mgh measured decrease fasting glucose compared pre treatment people risk factor type 2 diabetes\n",
      "0.36061415\n",
      "allele c associated increased response rosiglitazone people diabetes mellitus compared allele cyp2c8 slco1b1 variant therapeutic response thiazolidinediones patient type 2 diabetes measured hba1c reduction\n",
      "0.4068891\n",
      "allele associated increased response glipizide compared allele c tcf7l2 genetic variation augments incretin resistance influence response sulfonylurea metformin study understand genetics acute response metformin glipizide human sugar mgh measured shorter time steeper slope trough glucose level people risk factor type 2 diabetes\n"
     ]
    }
   ],
   "source": [
    "# summary based BOW experiments\n",
    "summary_data = drug_info_df['summary'].values.tolist() \n",
    "summary_dict = corpora.Dictionary(d.split() for d in summary_data)\n",
    "f2_cnt = len(summary_dict.token2id)\n",
    "summary_corpus = [summary_dict.doc2bow(row.split()) for row in summary_data]\n",
    "tfidf_summary = models.TfidfModel(summary_corpus)\n",
    "summary_index = similarities.SparseMatrixSimilarity(tfidf_summary[summary_corpus], num_features = f2_cnt)\n",
    "    \n",
    "text1 = pat_prof_df['generic_info'][2]\n",
    "text2 = pat_prof_df['diagnostic_info'][2]\n",
    "\n",
    "new_vec = summary_dict.doc2bow(text2.split())\n",
    "\n",
    "print(text2)\n",
    "sim = summary_index[tfidf_summary[new_vec]]\n",
    "i_2 = np.argpartition(sim, -15)[-15:] \n",
    "for ind_ in i_2:\n",
    "    print(sim[ind_])\n",
    "    print(drug_info_df['summary'][ind_])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
